packages/apollo-federation/src/composition/validate/sdl/uniqueFieldDefinitionNames.ts

```diff
@@ -0,0 +1,180 @@
+  | InterfaceTypeDefinitionNode
+  | ObjectTypeDefinitionNode;
+
+type TypeExtensionsRequiringUniqueFields =
+  | InputObjectTypeExtensionNode
+  | InterfaceTypeExtensionNode
+  | ObjectTypeExtensionNode;
+
+export function duplicateFieldDefinitionNameMessage(
+  typeName: string,
+  fieldName: string,
+): string {
+  return `Field "${typeName}.${fieldName}" can only be defined once.`;
+}
+
+export function existedFieldDefinitionNameMessage(
+  typeName: string,
+  fieldName: string,
+): string {
+  return `Field "${typeName}.${fieldName}" already exists in the schema. It cannot also be defined in this type extension.`;
+}
+
+/**
+ * Unique field definition names
+ *
+ * A GraphQL complex type is only valid if all its fields are uniquely named.
+ * Modified to permit duplicate field names on value types.
+ */
+export function UniqueFieldDefinitionNames(
+  context: SDLValidationContext,
+): ASTVisitor {
+  const schema = context.getSchema();
+  const existingTypeMap: TypeMap = schema
+    ? schema.getTypeMap()
+    : Object.create(null);
+  interface FieldToNameNodeMap {
+    [fieldName: string]: NameNode;
+  }
+  const knownFieldNames: {
+    [typeName: string]: FieldToNameNodeMap;
+  } = Object.create(null);
+
+  const possibleValueTypes: {
+    [key: string]: NodeTypesRequiringUniqueFields | undefined;
+  } = Object.create(null);
+
+  // Maintain original functionality for type extensions, but substitute our
+  // more permissive validator for base types to allow value types
+  return {
+    InputObjectTypeExtension: checkFieldUniqueness,
+    InterfaceTypeExtension: checkFieldUniqueness,
+    ObjectTypeExtension: checkFieldUniqueness,
+    InputObjectTypeDefinition: checkFieldUniquenessExcludingValueTypes,
+    InterfaceTypeDefinition: checkFieldUniquenessExcludingValueTypes,
+    ObjectTypeDefinition: checkFieldUniquenessExcludingValueTypes,
+  };
+
+  function checkFieldUniqueness(node: TypeExtensionsRequiringUniqueFields) {
+    const typeName = node.name.value;
+
+    if (!knownFieldNames[typeName]) {
+      knownFieldNames[typeName] = Object.create(null);
+    }
+
+    if (node.fields) {
+      const fieldNames = knownFieldNames[typeName];
+
+      for (const fieldDef of node.fields) {
+        const fieldName = fieldDef.name.value;
+
+        if (hasField(existingTypeMap[typeName], fieldName)) {
+          context.reportError(
+            new GraphQLError(
+              existedFieldDefinitionNameMessage(typeName, fieldName),
+              fieldDef.name,
+            ),
+          );
+        } else if (fieldNames[fieldName]) {
+          context.reportError(
+            new GraphQLError(
+              duplicateFieldDefinitionNameMessage(typeName, fieldName),
+              [fieldNames[fieldName], fieldDef.name],
+            ),
+          );
+        } else {
+          fieldNames[fieldName] = fieldDef.name;
+        }
+      }
+    }
+
+    return false;
+  }
+
+  function checkFieldUniquenessExcludingValueTypes(
```

Comment:
This logic seems right to me. It'd be easier to follow in the future if there were a few comments :)

Reply:
Resolved in b3a34ada